File: src/0429.n-ary-tree-level-order-traversal.764/solution.ts

function levelOrder (root: Node | null): number[][] {
    if (!root) {
        return [];
    }
    const result:number[][] = [];
    traversal([root, ], result);
    return result;
}

function traversal (levelNodes:Node[], result:number[][]) {
    const vals:number[] = [];
    const nextLevelNodes:Node[] = [];
    for (let i = 0; i < levelNodes.length; i++) {
        vals.push(levelNodes[i].val);
        nextLevelNodes.push(...levelNodes[i].children);
    }
    result.push(vals);
    nextLevelNodes.length && traversal(nextLevelNodes, result);
}

File: src/0429.n-ary-tree-level-order-traversal.764/solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> result = new ArrayList<>();
        if(root == null){
            return result;
        }
        levelOrder(List.of(root),result);
        return result;
    }

    private void levelOrder(List<Node> nodes,List<List<Integer>> result){
        List<Integer> row = new ArrayList<>();
        List<Node> nextRow = new ArrayList<>();
        for(Node node : nodes){
            row.add(node.val);
            nextRow.addAll(node.children);
        }
        result.add(row);
        if(!nextRow.isEmpty()){
            levelOrder(nextRow,result);
        }

    }

}